File: app/models/api/scraper_configs.py
"""API schemas for scraper configuration endpoints."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict

from app.models.api.base import UTCDateTime
from app.models.contracts import FeedSubscriptionOutcome


class ScraperConfigStatsResponse(BaseModel):
    """Derived stats for a single scraper configuration."""

    total_count: int
    completed_count: int
    unread_count: int
    processing_count: int
    latest_processed_at: UTCDateTime | None = None
    latest_publication_at: UTCDateTime | None = None
    next_expected_at: UTCDateTime | None = None
    average_interval_hours: float | None = None
    interval_sample_size: int = 0


class ScraperConfigResponse(BaseModel):
    """Response model for scraper configs."""

    id: int
    scraper_type: str
    display_name: str | None = None
    config: dict[str, Any]
    feed_url: str | None = None
    limit: int | None = None
    is_active: bool
    created_at: UTCDateTime
    stats: ScraperConfigStatsResponse | None = None
    subscription_outcome: FeedSubscriptionOutcome | None = None
    backfill_task_id: int | None = None

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": 1,
                "scraper_type": "substack",
                "display_name": "Import AI",
                "config": {"feed_url": "https://example.substack.com/feed"},
                "feed_url": "https://example.substack.com/feed",
                "limit": 10,
                "is_active": True,
                "created_at": "2025-06-24T12:00:00Z",
                "stats": {
                    "total_count": 8,
                    "completed_count": 7,
                    "unread_count": 3,
                    "processing_count": 1,
                    "latest_processed_at": "2026-03-30T21:30:00Z",
                    "latest_publication_at": "2026-03-30T20:00:00Z",
                    "next_expected_at": "2026-04-01T08:00:00Z",
                    "average_interval_hours": 24.0,
                    "interval_sample_size": 3,
                },
            }
        }
    )


class SubscribeToFeedRequest(BaseModel):
    """Request to subscribe to a detected feed."""

    feed_url: str
    feed_type: str
    display_name: str | None = None

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "feed_url": "https://example.substack.com/feed",
                "feed_type": "substack",
                "display_name": "Example Newsletter",
            }
        }
    )

File: rust/crates/newsly-contracts/src/scraper_configs.rs
use chrono::{DateTime, Utc};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use utoipa::ToSchema;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum ScraperType {
    Substack,
    Atom,
    PodcastRss,
    Youtube,
    Reddit,
    Aggregator,
}

impl ScraperType {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Substack => "substack",
            Self::Atom => "atom",
            Self::PodcastRss => "podcast_rss",
            Self::Youtube => "youtube",
            Self::Reddit => "reddit",
            Self::Aggregator => "aggregator",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Deserialize, JsonSchema, ToSchema)]
pub struct CreateUserScraperConfig {
    pub scraper_type: ScraperType,
    pub display_name: Option<String>,
    #[serde(default)]
    pub config: Map<String, Value>,
    #[serde(default = "default_active")]
    pub is_active: bool,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, JsonSchema, ToSchema)]
pub struct UpdateUserScraperConfig {
    pub display_name: Option<String>,
    pub config: Option<Map<String, Value>>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct SubscribeToFeedRequest {
    pub feed_url: String,
    pub feed_type: String,
    pub display_name: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum FeedSubscriptionOutcome {
    Created,
    Reactivated,
    AlreadySubscribed,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct ScraperConfigStatsResponse {
    pub total_count: i64,
    pub completed_count: i64,
    pub unread_count: i64,
    pub processing_count: i64,
    #[schemars(with = "Option<String>")]
    #[schema(value_type = Option<String>, format = DateTime)]
    pub latest_processed_at: Option<DateTime<Utc>>,
    #[schemars(with = "Option<String>")]
    #[schema(value_type = Option<String>, format = DateTime)]
    pub latest_publication_at: Option<DateTime<Utc>>,
    #[schemars(with = "Option<String>")]
    #[schema(value_type = Option<String>, format = DateTime)]
    pub next_expected_at: Option<DateTime<Utc>>,
    pub average_interval_hours: Option<f64>,
    #[serde(default)]
    pub interval_sample_size: usize,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct ScraperConfigResponse {
    pub id: i64,
    pub scraper_type: String,
    pub display_name: Option<String>,
    pub config: Map<String, Value>,
    pub feed_url: Option<String>,
    pub limit: Option<i64>,
    pub is_active: bool,
    #[schemars(with = "String")]
    #[schema(value_type = String, format = DateTime)]
    pub created_at: DateTime<Utc>,
    pub stats: Option<ScraperConfigStatsResponse>,
    pub subscription_outcome: Option<FeedSubscriptionOutcome>,
    pub backfill_task_id: Option<i64>,
}

const fn default_active() -> bool {
    true
}
